lib/number.js

```diff
@@ -0,0 +1,318 @@
+import {
+	getDocumentLocaleSettings,
+	getLanguage,
+	merge,
+	validateFormatValue
+} from './common.js';
+
+function formatPositiveInteger(value, descriptor) {
+	value = Math.floor(value);
+
+	const valueStr = '' + value;
+	let ret = '';
+
+	const groupSizes = [3];
+	let currentGroupSizeIndex = -1;
+	const maxGroupSizeIndex = groupSizes.length - 1;
+	let currentGroupSize = 0;
+	let groupEnd = valueStr.length;
+
+	while (groupEnd > 0) {
+		if (currentGroupSizeIndex < maxGroupSizeIndex) {
+			currentGroupSize = groupSizes[++currentGroupSizeIndex];
+		}
+
+		let chunk = null;
+		if (currentGroupSize === 0) {
+			chunk = valueStr.substring(0, groupEnd);
+		} else {
+			const groupStart = groupEnd - currentGroupSize;
+			chunk = valueStr.substring(groupStart, groupEnd);
+		}
+
+		// not first or only chunk
+		if (groupEnd !== valueStr.length) {
+			ret = descriptor.symbols.group + ret;
+		}
+
+		ret = chunk + ret;
+
+		groupEnd -= chunk.length;
+	}
+
+	return ret;
+}
+
+function formatExponentialDecimal(value) {
+	// Get a value should be like "1.25e-8"
+	const pieces = value.split('e-');
+	let ret = pieces[0].replace('.', '');
+	const zeroCount = parseInt(pieces[1]) - 1;
+
+	for (let i = 0; i < zeroCount; i++) {
+		ret = '0' + ret;
+	}
+	return ret;
+}
+
+function validateFormatOptions(options) {
+
+	options = options || {};
+
+	if (options.style !== 'decimal' && options.style !== 'percent') {
+		options.style = 'decimal';
+	}
+	options.minimumFractionDigits = validateInteger(
+		'minimumFractionDigits',
+		options.minimumFractionDigits,
+		0,
+		0,
+		20
+	);
+	options.maximumFractionDigits = validateInteger(
+		'maximumFractionDigits',
+		options.maximumFractionDigits,
+		Math.max(options.minimumFractionDigits, 3),
+		0,
+		20
+	);
+
+	if (options.minimumFractionDigits > options.maximumFractionDigits) {
+		throw new RangeError('maximumFractionDigits value is out of range.');
+	}
+
+	return options;
+
+}
+
+function validateInteger(name, value, defaultValue, min, max) {
+
+	if (value === undefined || value === null) {
+		value = defaultValue;
+	}
+	if (typeof value === 'string') {
+		value = parseInt(value);
+	}
+	if (isNaN(value) || typeof value !== 'number' || (min !== undefined && value < min) || (max !== undefined && value > max)) {
+		throw new RangeError(name + ' value is out of range.');
+	}
+
+	return value;
+
+}
+
+export function getNumberDescriptor() {
+
+	const language = getLanguage();
+
+	let negativePattern = '-{number}';
+	if (language === 'ar-sa') {
```

Comment:
it looks like this also applies to just `ar`

Reply:
Hmm yeah good call. Actually looking through here, I think I should be applying almost all of these number format settings based on the base locale and not the full code. Will fix.

Reply:
I've fixed this (and a couple other issues) and added tests for all the base languages.